service/history/execution/context.go

```diff
@@ -263,22 +264,27 @@ func (c *contextImpl) LoadWorkflowExecutionWithTaskVersion(
 	}
 
 	if c.mutableState == nil {
-		response, err := c.getWorkflowExecutionWithRetry(ctx, &persistence.GetWorkflowExecutionRequest{
-			DomainID:   c.domainID,
-			Execution:  c.workflowExecution,
-			DomainName: domainEntry.GetInfo().Name,
-		})
-		if err != nil {
-			return nil, err
-		}
+		var response *persistence.GetWorkflowExecutionResponse
+		for i := 0; i < checksumErrorRetryCount; i++ {
+			response, err = c.getWorkflowExecutionWithRetry(ctx, &persistence.GetWorkflowExecutionRequest{
+				DomainID:   c.domainID,
+				Execution:  c.workflowExecution,
+				DomainName: domainEntry.GetInfo().Name,
+			})
+			if err != nil {
+				return nil, err
+			}
 
-		c.mutableState = NewMutableStateBuilder(
-			c.shard,
-			c.logger,
-			domainEntry,
-		)
+			c.mutableState = NewMutableStateBuilder(
+				c.shard,
+				c.logger,
+				domainEntry,
+			)
 
-		c.mutableState.Load(response.State)
+			if err := c.mutableState.Load(response.State); err == nil {
```

Comment:
let's check err type and retry if it's checksum failure. I know that's only case `Load` will return err today but better to structure the code to be future proof.